openeo_pg_parser_networkx/resolving_utils.py

```diff
@@ -0,0 +1,217 @@
+    _remove_non_root_result(process_graph, root_result)
+    return process_graph
+
+
+'''
+Finds and resolves any sub process graphs within nodes like apply and reduce.
+'''
+
+
+def _resolve_sub_process_graphs(
+    process_graph: dict,
+    process_registry: ProcessRegistry,
+    get_udp_spec: Callable[[str], dict] = None,
+):
+    for _, node in process_graph.items():
+        if 'process' in node['arguments']:
+            node['arguments']['process']['process_graph'] = resolve_process_graph(
+                node['arguments']['process']['process_graph'],
+                process_registry,
+                get_udp_spec,
+            )
+    return process_graph
+
+
+def _fill_in_processes(
+    process_graph: dict[str, Any],
+    process_registry: ProcessRegistry,
+    get_udp_spec: Callable[[str], dict] = None,
+):
+    """
+    Recursively fill in UDPs with their respective definitions until
+    only predefined processes remain.
+
+    Ignores subgraphs, those get resolved individually.
+    """
+    for process_replacement_id, process in process_graph.items():
+        process_id = process['process_id']
+
+        if ('predefined', process_id) not in process_registry:
+            if ('user', process_id) not in process_registry and get_udp_spec is not None:
+                process_registry[('user', process_id)] = Process(
+                    get_udp_spec(process_id), implementation=None, namespace='user'
+                )
+
+            process_graph[process_replacement_id] = copy.deepcopy(
+                process_registry[('user', process_id)].spec['process_graph']
+            )
+
+            _remap_names(
+                process_graph=process_graph, process_replacement_id=process_replacement_id
+            )
+            _adjust_parameters(
+                process_graph=process_graph,
+                process_replacement_id=process_replacement_id,
+                arguments=process['arguments'],
+            )
+            _fill_in_processes(
+                process_graph=process_graph[process_replacement_id],
+                process_registry=process_registry,
+                get_udp_spec=get_udp_spec,
+            )
+
+
+def _remap_names(process_graph, process_replacement_id):
+    '''
+    Renames process nodes so that uniqueness is guaranteed, new names
+    follow the schema root_subnode_subsubnode etc.
+    '''
+    name_remapping = [
+        (f"{process_replacement_id}_{key}", key)
+        for key, _ in process_graph[process_replacement_id].items()
+    ]
+
+    for new_key, old_key in name_remapping:
+        process_graph[process_replacement_id][new_key] = process_graph[
+            process_replacement_id
+        ].pop(old_key)
+
+
+def _adjust_parameters(process_graph, process_replacement_id, arguments):
+    '''
+    "Rewires" the from_parameter connections of filled in UDPs to be consistent with
+    the parameters available from parent nodes.
+    '''
+    if len(arguments) > 0:
```

Comment:
Lmk what you think, but as far as I understand this, the arguments are only considered from the immediate parent here. According to https://openeo.org/documentation/1.0/developers/api/reference.html#section/Processes/Process-Graphs, ("Accessing process parameters"), parameter references can be resolved from the nearest parent that has it. Is this being addressed in the recursion somehow I'm not getting? Tbh I haven't encountered such a doubly nested process graph yet, but wanted your take on this regardless!

Reply:
I'm actually unsure if there is such a double nesting and how you would achieve it. You can doubly nest UDPs of course, but for those to be valid you'd have to wire the parameters for each level of nesting and then the recursion kicks in when resolving each level. 

So I don't know how a doubly nested UDP with a parameter that only gets used 2 steps down would even be possible to create.

Reply:
Right - doesn't seem very useful or likely to appear indeed, so probably not worth spending more time on it - thanks!